Make auto-advance flag a requirement for conditional/expression evaluation

SaracenOne opened the pull request:
This is a PR to address a potential major design discontinuity with Godot's StateMachine system. This has been bothering me for a while, and I believe I understand why certain aspects of AnimationStateMachine's feel so counter-inuitive. I feel it is a consequence of two distinct ways of traversing them: via travel or via auto-advance/conditions. With travel, conditions are actually ignored, and it is a regular A* traversal to simply get to a desired node, whereas auto-advance can use conditional branch to have a state machine essentially drive itself.

The problem I feel with the current implementation is that you have mainly three types of transitions:
- No auto-advance and no condition: does nothing on its own
- No auto-advance but has conditions: advances but if conditions are met
- Auto-advance: ignores all conditions and advances regardless

My problem is that auto-advance more or less overwrites the behaviour conditions, and this feels very strange from a usability standpoint. It essentially makes the it an either/or scenario, and the relationship between these two elements is not obvious. My proposal therefore is to make it so that for any automatic traversal, auto-advance must always be on, and conditional evaluation is only done if auto-advance is set. While this PR only changes the actual behaviour and documentation to provide better explanation, I think more could be done on the editor side to make the relationship between the two more obvious. I could also supplement this with features such as allowing you to set the default state of auto-advance on new transitions in the same way you can for switch mode.

I feel strongly on the whole that short of removing auto-advance entirely (which doesn't feel fair given not having auto-advance active can make it useful for the 'travel' method), this is the best way to address a pretty glaring design issue, and it feels relevant to address this given the potential that a lot of game logic can be implemented purely via state machines.

SaracenOne commented:
![godot windows tools x86_64_szWO18vJIt](https://user-images.githubusercontent.com/12756047/188301090-7c8f58ed-8fae-4522-ac92-54339f2edb32.png)
Added a new auto-advance toggle button next to the transition switch mode for creating new transitions. Feels important to do for the sake of workflow speed, though I do wonder if there's ways we could make the functionality more obvious.

SaracenOne commented:
I went and added the context-sensitive behaviour anyway. I could really see a reason not to do it given the way the state machine editor currently uses, and feel like it will make it easier for people to understand how it works.
![godot windows tools x86_64_0zuIXhKQ07](https://user-images.githubusercontent.com/12756047/188302127-2461e68c-85f2-4c19-a962-91919fe6e860.gif)

fire commented:
@TokageItLab what do you think?

akien-mga commented on scene/animation/animation_node_state_machine.cpp:
```diff
@@ -666,15 +666,17 @@ bool AnimationNodeStateMachinePlayback::_check_advance_condition(const Ref<Anima
-			if (!exp->has_execute_failed()) {
-				if (ret) {
-					return true;
+			if (exp->has_execute_failed()) {
+				return false;
+			} else {
+				if (!ret) {
+					return false;
 				}
 			}
```
Could be:
```
if (exp->has_execute_failed() || !ret) { return false; }
```
I guess, unless we plan to add more stuff in the `else`.

akien-mga commented on editor/plugins/animation_state_machine_editor.cpp:
```diff
@@ -1941,35 +1950,51 @@ AnimationNodeStateMachineEditor::AnimationNodeStateMachineEditor() {
-	tool_erase_hb->add_child(tool_erase);
+	selection_tools_hb->add_child(tool_erase);
+
+	// Context sensitive transition tools:
```
"Context-sensitive" used a hyphen above.

akien-mga commented on doc/classes/AnimationNodeStateMachineTransition.xml:
```diff
@@ -9,7 +9,7 @@
 	<members>
 		<member name="advance_condition" type="StringName" setter="set_advance_condition" getter="get_advance_condition" default="&amp;&quot;&quot;">
-			Turn on auto advance when this condition is set. The provided name will become a boolean parameter on the [AnimationTree] that can be controlled from code (see [url=$DOCS_URL/tutorials/animation/animation_tree.html#controlling-from-code]Using AnimationTree[/url]). For example, if [member AnimationTree.tree_root] is an [AnimationNodeStateMachine] and [member advance_condition] is set to [code]" ...
+			Will advance to the next state if [member auto_advance] is true and the condition is set. The provided name will become a boolean parameter on the [AnimationTree] that can be controlled from code (see [url=$DOCS_URL/tutorials/animation/animation_tree.html#controlling-from-code]Using AnimationTree[/url]). For example, if [member AnimationTree.tree_root] is an [AnimationNodeStateMachine] and [me ...
```
We style booleans as `[code]true[/code]` and `[code]false[/code]` in the docs (same below).

TokageItLab commented:
Does this PR solve this issue? 
https://github.com/godotengine/godot-proposals/issues/2200

SaracenOne commented:
@TokageItLab It doesn't, but I've started experimenting with transition interrupts as part of major evaluation of state machine usability. I'm hoping I can add that as a new feature soon, along with more granular support of the exact timing of when transitions should occur.

SaracenOne committed:
Change auto_advance flag to advance_mode enum and unify with disabled flag.
Expressions and conditions now require auto mode to be set to auto.
Adds a toggle button to the state machine editor for whether new transitions
advance settings should default to auto mode or not.

reduz commented:
I think I recall we discussed using an enum for this, what happened in the end?

mhilbrunner commented:
Animation meeting: change to use an enum, maybe add `disable`. @SaracenOne wants to get this done in the next few days.

SaracenOne commented:
This has been updated with the change to a single enum. I sincerely apologize for not following up on this sooner.

akien-mga commented:
Thanks!

TokageItLab commented on scene/animation/animation_node_state_machine.cpp:
```diff
@@ -628,14 +618,14 @@ double AnimationNodeStateMachinePlayback::process(AnimationNodeStateMachine *p_s
-		return true;
+	if (transition->get_advance_mode() != AnimationNodeStateMachineTransition::ADVANCE_MODE_AUTO) {
+		return false;
 	}
 
 	StringName advance_condition_name = transition->get_advance_condition_name();
 
-	if (advance_condition_name != StringName() && bool(state_machine->get_parameter(advance_condition_name))) {
-		return true;
+	if (advance_condition_name != StringName() && !bool(state_machine->get_parameter(advance_condition_name))) {
+		return false;
```
Is this change in decision correct?

SaracenOne replied:
I'm reading this over, and unless I've got something very wrong, it should be. The first check is to return false if the advance mode is not set to auto since only auto advance evaluates conditions and expressions (though this should probably instead be changed to ```transition->get_advance_mode() == AnimationNodeStateMachineTransition::ADVANCE_MODE_DISABLED)``` in case we want to add condition evaluation to the travel function in the future. As for the second check, this should also be correct in that first we're checking if the advance condition string is set to anything, then evaluating it and return false if the parameter is also false.
